File: JUnit2/Example-6/src/test/java/edu/odu/cs/cs350/examples/TestRoster.java
package edu.odu.cs.cs350.examples;

import org.junit.FixMethodOrder;
import org.junit.runners.MethodSorters;
import org.junit.Test;
import org.junit.Before;

import static org.junit.Assert.*;

import static org.hamcrest.CoreMatchers.*;
import org.hamcrest.core.IsNull;

/**
 * 1 - Does this piece of code perform the operations
 *     it was designed to perform?
 *
 * 2 - Does this piece of code do something it was not
 *     designed to perform?
 *
 * 1 Test per mutator
 */
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestRoster {
    // Students - will not be changed during tests
    Student john  = new Student("John");
    Student tom   = new Student("Tom");
    Student jay   = new Student("Jay");
    Student oscar = new Student("Oscar");

    // Student "lists" - will not be changed during tests
    Student[] firstThreeStudents = {john, tom, jay};
    Student[] allStudents        = {john, tom, jay, oscar};

    // Course Rosters - will possibly change in each test
    Roster defaultCourse;
    Roster emptyCS350;

    @Before
    public void setUp()
    {
        defaultCourse = new Roster();
        emptyCS350    = new Roster(3, "CS 350");
    }

    @Test
    public void testDefaultConstructor()
    {
        assertThat(defaultCourse.getCourseNum(), equalTo("CS 150"));
        assertThat(defaultCourse.getEnrollLimit(),
                   equalTo(Roster.DEFAULT_MAX_STUDENTS));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(defaultCourse.listEnrolledStudents(), notNullValue());
        assertThat(defaultCourse.listEnrolledStudents().size(), equalTo(0));

        // skipping hashcode
        // skipping equals

        // test toString
        assertThat(defaultCourse.toString(), containsString("CS 150"));
        assertThat(defaultCourse.toString(),
                   containsString(Integer.toString(defaultCourse.numEnrolled())));
        assertThat(defaultCourse.toString(),
                   containsString(Integer.toString(Roster.DEFAULT_MAX_STUDENTS)));
    }

    @Test
    public void testNonDefaultConstructor()
    {
        assertThat(emptyCS350.getCourseNum(), equalTo("CS 350")); // fixed mistake
        assertThat(emptyCS350.getEnrollLimit(), equalTo(3));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(emptyCS350.listEnrolledStudents(), notNullValue());
        assertThat(emptyCS350.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(emptyCS350.hashCode(), not(equalTo(defaultCourse.hashCode())));
        // NOT skipping equals
        assertThat(emptyCS350, not(equalTo(defaultCourse)));

        // test toString
        assertThat(emptyCS350.toString(), containsString("CS 350"));
        assertThat(emptyCS350.toString(),
                   containsString(Integer.toString(emptyCS350.numEnrolled())));
        assertThat(emptyCS350.toString(), containsString(Integer.toString(3)));
    }

    @Test
    public void testSetCourseNum()
    {
        Roster cs252 = new Roster();

        int oldHashCode = cs252.hashCode();

        cs252.setCourseNum("CS 252");

        assertThat(cs252.getCourseNum(), equalTo("CS 252"));
        assertThat(cs252.getEnrollLimit(),
                   equalTo(Roster.DEFAULT_MAX_STUDENTS));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(cs252.listEnrolledStudents(), notNullValue());
        assertThat(cs252.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(cs252.hashCode(), not(equalTo(oldHashCode)));
        // NOT skipping equals
        assertThat(cs252, not(equalTo(defaultCourse)));

        // test toString
        assertThat(cs252.toString(), containsString("CS 252"));
        assertThat(cs252.toString(),
                   containsString(Integer.toString(cs252.numEnrolled()))); // fixed mistake
        assertThat(cs252.toString(),
                   containsString(Integer.toString(Roster.DEFAULT_MAX_STUDENTS)));
    }

    @Test
    public void testSetEnrollLimit()
    {
        assertThat(emptyCS350.getCourseNum(), equalTo("CS 350"));

        emptyCS350.setEnrollLimit(2);

        assertThat(emptyCS350.getEnrollLimit(), equalTo(2));

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // No students have been added
        assertThat(emptyCS350.listEnrolledStudents(), notNullValue());
        assertThat(emptyCS350.listEnrolledStudents().size(), equalTo(0));

        // NOT skipping hashcode
        assertThat(emptyCS350.hashCode(), not(equalTo(defaultCourse.hashCode())));
        // NOT skipping equals
        assertThat(emptyCS350, not(equalTo(defaultCourse)));

        // test toString
        assertThat(emptyCS350.toString(), containsString("CS 350"));
        assertThat(emptyCS350.toString(),
                   containsString(Integer.toString(emptyCS350.numEnrolled())));
        assertThat(emptyCS350.toString(), containsString(Integer.toString(2)));
    }

    @Test
    public void testEnroll()
    {
        //This is where the fun starts
        Roster cs725 = new Roster(3, "CS 725");

        int oldHashCode = cs725.hashCode();

        assertThat(cs725.getCourseNum(), equalTo("CS 725"));
        assertThat(cs725.getEnrollLimit(), equalTo(3)); // fixed mistake

        assertThat(defaultCourse.numEnrolled(), equalTo(0));

        // try to add 4 students
        assertThat(cs725.enroll(john), is(true));
        assertThat(cs725.enroll(tom), is(true));
        assertThat(cs725.enroll(jay), is(true));
        assertThat(cs725.enroll(oscar), is(false)); // should fail (limit of 3)

        // 3 students have been added
        assertThat(cs725.listEnrolledStudents(), notNullValue());
        assertThat(cs725.listEnrolledStudents().size(), equalTo(3));
        assertThat(cs725.listEnrolledStudents().toArray(),
                   equalTo(firstThreeStudents));

        oldHashCode = cs725.hashCode();

        // Change the limit to 4
        emptyCS350.setEnrollLimit(4);
        assertThat(emptyCS350.getEnrollLimit(), equalTo(4));

        // try to add a 4th student with the new limit of 4
        assertThat(cs725.enroll(oscar), is(true)); // should succeed (limit of 4)

        // 4 students have been added
        assertThat(cs725.listEnrolledStudents(), notNullValue());
        assertThat(cs725.listEnrolledStudents().size(), equalTo(4));
        assertThat(cs725.listEnrolledStudents().toArray(),
                   equalTo(firstThreeStudents));

        // NOT skipping hashcode
        assertThat(cs725.hashCode(), not(equalTo(oldHashCode)));
        // NOT skipping equals
        assertThat(cs725, not(equalTo(defaultCourse)));

        // **test flaw** - did not exercise adding the same student twice
        assertThat(cs725.enroll(tom), is(false)); // should fail

        // test toString
        assertThat(cs725.toString(), containsString("CS 725"));
        assertThat(cs725.toString(),
                   containsString(Integer.toString(cs725.numEnrolled())));
        assertThat(cs725.toString(),
                   containsString(Integer.toString(3)));

        assertThat(cs725.toString(), containsString(allStudents[0].toString()));
        assertThat(cs725.toString(), containsString(allStudents[1].toString()));
        assertThat(cs725.toString(), containsString(allStudents[2].toString()));
        assertThat(cs725.toString(), containsString(allStudents[3].toString()));

    }

    @Test
    public void testClone()
    {
        fail("Left as an Exercise");
    }
}

File: Office-Hours/JSON-Rust/tests/test_roster.rs
use std::collections::HashSet;

use hamcrest2::prelude::*;
use rstest::fixture;
use rstest::rstest;

use enroll_students::prelude::*;
use enroll_students::roster::DEFAULT_MAX_STUDENTS;

#[fixture]
fn default_courses() -> (Roster, Roster) {
    let default_course = Roster::default();
    let empty_cs350 = Roster::new(3, "CS 350");

    (default_course, empty_cs350)
}

type StudentListTuple = (
    Student,
    Student,
    Student,
    Student,
    [Student; 4],
    [Student; 3],
);

#[fixture]
fn students_and_lists() -> StudentListTuple {
    let john = Student::new("John");
    let tom = Student::new("Tom");
    let jay = Student::new("Jay");
    let oscar = Student::new("Oscar");

    // Student "lists" - will not be changed during tests
    let all_students = [john.clone(), tom.clone(), jay.clone(), oscar.clone()];
    let first_three_students = [john.clone(), tom.clone(), jay.clone()];

    (john, tom, jay, oscar, all_students, first_three_students)
}

#[rstest]
fn test_default_constructor() {
    let default_course = Roster::default();

    assert_that!(&default_course.course_num, equal_to("CS 150"));
    assert_that!(default_course.enroll_limit, equal_to(DEFAULT_MAX_STUDENTS));

    assert_that!(default_course.num_enrolled(), equal_to(0_usize));

    // No students have been added
    assert_that!(
        default_course.list_enrolled_students().len(),
        equal_to(0_usize)
    );

    // skipping hashcode
    // skipping equals

    // test __str__
    let default_course_str = default_course.to_string();
    assert_that!(default_course_str.contains("CS 150"), is(true));
    assert_that!(
        default_course_str.contains(&default_course.num_enrolled().to_string()),
        is(true)
    );
    assert_that!(
        default_course_str.contains(&DEFAULT_MAX_STUDENTS.to_string()),
        is(true)
    );
}

#[rstest]
fn test_non_default_constructor(default_courses: (Roster, Roster)) {
    let (default_course, empty_cs350) = default_courses;

    assert_that!(&empty_cs350.course_num, equal_to("CS 350"));
    assert_that!(empty_cs350.enroll_limit, equal_to(3));

    assert_that!(default_course.num_enrolled(), equal_to(0));

    // No students have been added
    // assert_that!(empty_cs350.list_enrolled_students(), is(none()));
    assert_that!(empty_cs350.list_enrolled_students().len(), equal_to(0));

    // NOT skipping hashcode
    // assert_that!(hash(empty_cs350), is_not(equal_to(hash(default_course))))
    // NOT skipping equals
    assert_that!(&empty_cs350, is_not(equal_to(&default_course)));

    // test __str__
    let empty_cs350_str = empty_cs350.to_string();
    assert!(empty_cs350_str.contains("CS 350"));
    assert!(empty_cs350_str.contains("3"));
}

#[rstest]
fn test_set_course_num(default_courses: (Roster, Roster)) {
    let (default_course, _) = default_courses;

    let mut cs252 = Roster::default();

    // let old_hash_code = cs252.hash();

    cs252.course_num = "CS 252".into();

    assert_that!(&cs252.course_num, equal_to("CS 252"));
    assert_that!(cs252.enroll_limit, equal_to(DEFAULT_MAX_STUDENTS));

    assert_that!(default_course.num_enrolled(), equal_to(0));

    // No students have been added
    assert_that!(cs252.list_enrolled_students().len(), equal_to(0));

    // NOT skipping hashcode
    // assert_that!(hash(cs252), is_not(equal_to(old_hash_code)));
    // NOT skipping equals
    assert_that!(&cs252, is(not(equal_to(&default_course))));

    // test __str__
    let cs252_str = cs252.to_string();
    assert!(cs252_str.contains("CS 252"));
    assert!(cs252_str.contains(&format!("{}", cs252.num_enrolled())));
    assert!(cs252_str.contains(&format!("{}", DEFAULT_MAX_STUDENTS)));
}

#[rstest]
fn test_set_enroll_limit(default_courses: (Roster, Roster)) {
    let (default_course, mut empty_cs350) = default_courses;

    empty_cs350.enroll_limit = 2;

    assert_that!(&empty_cs350.course_num, equal_to("CS 350"));
    assert_that!(empty_cs350.enroll_limit, equal_to(2));

    assert_that!(default_course.num_enrolled(), equal_to(0));

    // No students have been added
    assert_that!(empty_cs350.list_enrolled_students().len(), equal_to(0));

    // NOT skipping hashcode
    // assert_that!(hash(empty_cs350), is_not(equal_to(hash(default_course))));
    // NOT skipping equals
    assert_that!(&empty_cs350, is_not(equal_to(&default_course)));

    // test __str__
    let empty_cs350_str = empty_cs350.to_string();
    assert!(empty_cs350_str.contains("CS 350"));
    assert!(empty_cs350_str.contains(&format!("{}", empty_cs350.num_enrolled())));
    assert!(empty_cs350_str.contains("2"));
}

#[rstest]
fn test_enroll(default_courses: (Roster, Roster), students_and_lists: StudentListTuple) {
    let (john, tom, jay, oscar, all_students, first_three_students) = students_and_lists;
    let (default_course, _) = default_courses;

    // This is where the fun starts
    let mut cs725 = Roster::new(3, "CS 725");

    // old_hash_code = hash(cs725)

    // try to add 4 students
    assert_that!(cs725.enroll(john.clone()), is(ok()));
    assert_that!(cs725.enroll(tom.clone()), is(ok()));
    assert_that!(cs725.enroll(jay.clone()), is(ok()));
    assert_that!(cs725.enroll(oscar.clone()), is(err())); // should fail (limit of 3);

    assert_that!(&cs725.course_num, equal_to("CS 725"));
    assert_that!(cs725.enroll_limit, equal_to(3));
    assert_that!(cs725.num_enrolled(), equal_to(3)); // fixed mistake

    // 3 students have been added
    // assert_that!(cs725.list_enrolled_students(), not_none());
    assert_that!(cs725.list_enrolled_students().len(), equal_to(3));
    assert_that!(
        cs725
            .list_enrolled_students()
            .difference(&HashSet::from(first_three_students))
            .count(),
        is(equal_to(0))
    );

    // assert_that!(hash(cs725), is_not(equal_to(old_hash_code)));

    // old_hash_code = hash(cs725);

    // Change the limit to 4
    cs725.enroll_limit = 4;
    assert_that!(cs725.enroll_limit, equal_to(4));

    // try to add a 4th student with the new limit of 4
    assert_that!(cs725.enroll(oscar), is(ok())); // should succeed (limit of 4);

    // 4 students have been added
    // assert_that!(cs725.list_enrolled_students(), not_none());
    assert_that!(cs725.list_enrolled_students().len(), equal_to(4));
    // assert_that!(
    // cs725.list_enrolled_students(),
    // contains_inanyorder(*all_students),
    // );

    // NOT skipping hashcode
    // assert_that!(hash(cs725), is(not(equal_to(old_hash_code))));
    // NOT skipping equals
    assert_that!(&cs725, is_not(equal_to(&default_course)));

    // **test flaw** - did not exercise adding the same student twice
    assert_that!(cs725.enroll(tom), is(err())); // should fail

    // test __str__
    assert!(cs725.to_string().contains("CS 725"));
    assert!(
        cs725
            .to_string()
            .contains(&format!("{}", cs725.num_enrolled()))
    );
    assert!(cs725.to_string().contains("4"));

    assert!(cs725.to_string().contains(&all_students[0].to_string()));
    assert!(cs725.to_string().contains(&all_students[1].to_string()));
    assert!(cs725.to_string().contains(&all_students[2].to_string()));
    assert!(cs725.to_string().contains(&all_students[3].to_string()));
}
